### PR TITLE
8314226: Series of colon-style fallthrough switch cases with guards compiled incorrectly

### PR DESCRIPTION
**Pull request by biboudis**
The `switch` structure is translated in `handleSwitch` where we rewrite pattern matching switches. In some occasions a switch has multiple cases with multiple patterns where the n-th case does not complete normally and falls through to the n+1-st case:

```
switch (obj) {
   case Integer _ when ((Integer) obj) > 0:
   case String _ when !((String) obj).isEmpty():
      return 1;
   default:
      return -1;
}
```

This PR addresses that by translating the second case correctly and also replicates the body of the latter to the former (which we can do because no bindings are introduced in either statement blocks by the definition of unnamed variables.).

Previously the code was translated into:

```
switch (java.lang.runtime.SwitchBootstraps.typeSwitch(selector0$temp, index$1)) {
    case 0:
        Integer _;
        if (!(true && ((Integer)obj) > 0)) {
            index$1 = 1;
            continue;
        }

    case 1 when !((String)obj).isEmpty():
        return 1;

    default:
        return -1;
}
```

This PR adjusts the translation into:

```
switch (java.lang.runtime.SwitchBootstraps.typeSwitch(selector0$temp, index$1)) {
case 0:
    Integer _;
    if (!(true && ((Integer)obj) > 0)) {
        index$1 = 1;
        continue;
    }
    return 1;

case 1:
    String _;
    if (!((selector0$temp instanceof String || false) && (true && !((String)obj).isEmpty()))) {
        index$1 = 2;
        continue;
    }
    return 1;
}
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8314226](https://bugs.openjdk.org/browse/JDK-8314226): Series of colon-style fallthrough switch cases with guards compiled incorrectly (**Bug** - P4)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to [c46952b3](https://git.openjdk.org/jdk/pull/15532/files/c46952b3dc9fad2cc37e14ddbb24e5831382817e)
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15532/head:pull/15532` \
`$ git checkout pull/15532`

Update a local copy of the PR: \
`$ git checkout pull/15532` \
`$ git pull https://git.openjdk.org/jdk.git pull/15532/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15532`

View PR using the GUI difftool: \
`$ git pr show -t 15532`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15532.diff">https://git.openjdk.org/jdk/pull/15532.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15532#issuecomment-1702534452)

### CODE REVIEW

**Comment by mcimadamore** [comment]
So, as we discussed offline, type tests are normally enforced in the switch bootstrap. The first time we hit the test, we know we have some object and that the first label expects an Integer, so we test for that. But if the `when` clause fails, javac tells us to try again from the second label. Now, while the proposed code to the generated code looks ok - I wonder if this condition shouldn't instead be detected by the switch bootstrap: e.g. we're asking it to start again from a `case` which expects a `String`, but we really have some other Object on our hands, so that label index should never be returned by the switch bootstrap, which should either pick another applicable label index (if it exists) or just pick the default.

This would allow us to keep the generated code as is, and avoid introducing coupling between cases in the generated code.

**Comment by mcimadamore** [comment]
> So, as we discussed offline, type tests are normally enforced in the switch bootstrap. The first time we hit the test, we know we have some object and that the first label expects an Integer, so we test for that. But if the `when` clause fails, javac tells us to try again from the second label. Now, while the proposed code to the generated code looks ok - I wonder if this condition shouldn't instead be detected by the switch bootstrap: e.g. we're asking it to start again from a `case` which expects a `String`, but we really have some other Object on our hands, so that label index should never be returned by the switch bootstrap, which should either pick another applicable label index (if it exists) or just pick the default.
> 
> This would allow us to keep the generated code as is, and avoid introducing coupling between cases in the generated code.

Btw, looking at the comment in SwitchBootstraps::createRepeatIndexSwitch, it seems that's already the case:

```
/*
     * Construct test chains for labels inside switch, to handle switch repeats:
     * switch (idx) {
     *     case 0 -> if (selector matches label[0]) return 0; else if (selector matches label[1]) return 1; else ...
     *     case 1 -> if (selector matches label[1]) return 1; else ...
     *     ...
     * }
     */
```

**Comment by mcimadamore** [comment]
> > So, as we discussed offline, type tests are normally enforced in the switch bootstrap. The first time we hit the test, we know we have some object and that the first label expects an Integer, so we test for that. But if the `when` clause fails, javac tells us to try again from the second label. Now, while the proposed code to the generated code looks ok - I wonder if this condition shouldn't instead be detected by the switch bootstrap: e.g. we're asking it to start again from a `case` which expects a `String`, but we really have some other Object on our hands, so that label index should never be returned by the switch bootstrap, which should either pick another applicable label index (if it exists) or just pick the default.
> > This would allow us to keep the generated code as is, and avoid introducing coupling between cases in the generated code.
> 
> Btw, looking at the comment in SwitchBootstraps::createRepeatIndexSwitch, it seems that's already the case:
> 
> ```
> /*
>      * Construct test chains for labels inside switch, to handle switch repeats:
>      * switch (idx) {
>      *     case 0 -> if (selector matches label[0]) return 0; else if (selector matches label[1]) return 1; else ...
>      *     case 1 -> if (selector matches label[1]) return 1; else ...
>      *     ...
>      * }
>      */
> ```

Also, in the original test here:
https://bugs.openjdk.org/browse/JDK-8314226

It seems to me that the problem is slightly different - e.g. the client is passing a String, but the `when` clause doesn't seem to be applied by javac. Which seems to conflict with the original generated code you report above: if javac currently generates this:

```
 case 1 when !((String)obj).isEmpty():
        return 1;
```

How can javac go ahead with this case, considering that the provided string (in the original test) is empty?

I wonder if the real issue is that there is no real `when` guard emitted for the second `case` (because of javac getting confused with cascading).

**Comment by mcimadamore** [comment]
> I wonder if the real issue is that there is no real `when` guard emitted for the second `case` (because of javac getting confused with cascading).

To be clear, IMHO the generated code should be:

```
switch (java.lang.runtime.SwitchBootstraps.typeSwitch(selector0$temp, index$1)) {
case 0:
    Integer _;
    if (!(true && ((Integer)obj) > 0)) {
        index$1 = 1;
        continue;
    }
    return 1;

case 1:
    String _;
    if (true && !((String)obj).isEmpty())) {
        index$1 = 2;
        continue;
    }
    return 1;
}
```

**Comment by biboudis** [comment]
I think you are right @mcimadamore. We do not need to generate the extraneous condition in the `if`. Removed and updated the PR.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransPatterns.java [hunk 1]
```diff
@@ -575,7 +575,8 @@ private void handleSwitch(JCTree tree,
                             }
                         }
 
-                        if (c.guard != null) {
+                        boolean hasGuard = c.guard != null;
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransPatterns.java:578]: Can this declaration occur before the enclosing `if` ?

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransPatterns.java [hunk 2]
```diff
@@ -585,13 +586,20 @@ private void handleSwitch(JCTree tree,
...
                         JCContinue continueSwitch = make.at(clearedPatterns.head.pos()).Continue(null);
                         continueSwitch.target = tree;
-                        c.stats = c.stats.prepend(make.If(makeUnary(Tag.NOT, test).setType(syms.booleanType),
-                                                           make.Block(0, List.of(make.Exec(make.Assign(make.Ident(index),
-                                                                                                       makeLit(syms.intType, i + labels.length()))
-                                                                                     .setType(syms.intType)),
-                                                                                 continueSwitch)),
-                                                           null));
+
+                        if (previousC != null && hasGuard && previousCompletesNormally) {
```
[the remark this answers is not in the document]
Comment (biboudis) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransPatterns.java]: Correct. I refactored and extracted the code in a separate step.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransPatterns.java (hunk 1 above)
Comment (biboudis) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransPatterns.java]: Extracted the inlined code in a separate step.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransPatterns.java [hunk 3]
```diff
@@ -670,6 +659,41 @@ private void handleSwitch(JCTree tree,
...
+    //     case 0:
+    //         if (!((Integer)obj) > 0) { ... }
+    //         return 1;
+    //     case 1:
+    //         if (!((String)obj).isEmpty()) { ... }
+    //         return 1;
+    //     ...
+    // }
+    private static List<JCCase> patchCompletingNormallyCases(List<JCCase> cases) {
+        ListBuffer<JCCase> newCases = new ListBuffer<>();
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransPatterns.java:682]: Do you need to create a new buffer? Isn't this method just side-effecting the various cases?

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransPatterns.java [hunk 4]
```diff
@@ -541,7 +543,7 @@ private void handleSwitch(JCTree tree,
                     validCaseLabelList = clearedPatterns.head.hasTag(Tag.PATTERNCASELABEL);
                 }
 
-                if (validCaseLabelList && !previousCompletesNormally) {
+                if (validCaseLabelList && !previousCompletesNormally || c.guard != null) {
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransPatterns.java:546]: I'd personally add parenthesis to make it obvious which test has priority

**Review by mcimadamore — APPROVED** [review]
Looks good!

**Comment by lahodaj** [comment]
I think I'd like to ask for some more tests. Especially:
- multiple cases with statements where the fall through goes from one to the other, e.g. something like:
```
   case Integer _ when ((Integer) obj) > 0:
   case String _ when !((String) obj).isEmpty():
       System.err.println(1);
   case null:
       System.err.println(2);
   default:    
       System.err.println(3);
```
- cases with loop statements with break (and/or continue) to the loop. E.g.:
```
while (true) {break;}
````
instead of `System.err.println(2);`

- break/continue, return, yield(?) going outside of the switch (probably one of them would be enough) in the duplicated section.

**Comment by biboudis** [comment]
Thank you Jan. Does `multipleCasesWithReturn` cover your third bullet? Did I understood correctly?

**Comment by biboudis** [comment]
Thank you @lahodaj for the improvement. This is extremely elegant and reduces the code size!!!

**Review by vicente-romero-oracle — APPROVED** [review]
looks sensible

**Pull request closed without merging**